codalab/apps/common/competition_utils.py

```diff
@@ -15,12 +15,14 @@
 
 def get_most_popular_competitions(min_participants=400, limit=5, fill_in=True):
     today = datetime.datetime.today()
+    # TODO: This queryset returns none when `.order_by('?')` is used, breaking this functionality.
     competitions = Competition.objects.filter(published=True) \
         .filter(Q(end_date__gte=today) | Q(end_date=None)) \
         .annotate(num_participants=Count('participants')) \
         .filter(num_participants__gte=min_participants) \
-        .order_by('?') \
         .select_related('creator')[:limit]
+        # .order_by('?') \
+        # .select_related('creator')[:limit]
```

Comment:
so this fixes a test but what are the consequences?

Reply:
Previously that queryset was returning 0 competitions no matter what. While I was debugging it I removed several lines to see which one was causing it, and this was the only line causing no competitions to be returned.

If I understand correctly, `.order_by('?')` sorts them randomly. I guess in this case it'll have only the oldest competitions if there's more than the limit, which is less than ideal for this feature.

I'll work on this more and see if I can come up with a better solution.

Reply:
I think I found a solution, but it's not the best looking.

According to https://code.djangoproject.com/ticket/26390 the issue is with using `annotate` and `.order_by('?')`. Someone posted a work-around, and this is what it would look like implemented:

```
    competitions = Competition.objects.filter(published=True) \
                       .filter(Q(end_date__gte=today) | Q(end_date=None)) \
                       .filter(pk__in=Competition.objects \
                               .annotate(num_participants=Count('participants')) \
                               .filter(num_participants__gte=min_participants) \
                               ) \
                       .order_by('?') \
                       .select_related('creator')[:limit]
```

Tests pass with the work-around, but I think we'll be adding another query with it. I'm also not sure what would be preferred on the styling. If this looks like a good solution I'll push it.

Reply:
Thought I wrote a comment to this, woops. Yeah please try to refactor this to look a bit nicer, break out the sub query for example into another var

Reply:
You did have another comment that looks like it was marked resolved. The changes have been made as well. Is it possible you were looking at a specific commit?

![image](https://user-images.githubusercontent.com/28552312/88085161-543c4880-cb3a-11ea-88e7-415c9c299683.png)

Reply:
Huh? yeah commented elsewhere, closing this